### ISSUE TITLE
Add silent mode for console command

### ISSUE THREAD

**Issue by iRipVanWinkle** [yiisoft/yii2#13040]
### What steps will reproduce the problem?
I want to run console command or use it class (as migration) without output information in console.

When I use the command in the console it is not critical. But if I use the command class in my project and don't want it to be displayed messages to the console. I can't do it.

An example: `yii migrate --silent`

**Comment by SilverFire** [comment]
Might be implemented. As a workaround, you can run `yii migrate > /dev/null`

**Comment by klimov-paul** [comment]
Use
```
yii migrate --interactive=0
```

**Comment by SilverFire** [comment]
`yii migrate --interactive=0 > /dev/null`

**Comment by klimov-paul** [comment]
Yes that is correct.

**Comment by iRipVanWinkle** [comment]
But, If I want use:
```php
$migrateController = new MigrateController('migrate', $app);
$migrateController->migrationPath = $migrationPath;
$migrateController->runAction($command, ['interactive' => 0]);
```

**Comment by klimov-paul** [comment]
https://github.com/yiisoft/yii2/issues/1772#issuecomment-31605050

**Comment by iRipVanWinkle** [comment]
This solves the output `echo` but not the decides output `php://stdout`

**Comment by cebe** [comment]
The more generic way to implement this would be to allow redirection of output by overriding `console\Controller::stdout()` and `stderr()` methods. We could add configuration params to Controller to allow setting the streams to use for that. To pipe all output to nowhere you can set it to use `fopen('php://memory')`.

**Comment by iRipVanWinkle** [comment]
@cebe if need I can implement it? Because I decided my problem, but maybe interesting  it someone else.

**Comment by klimov-paul** [comment]
Relates #12862

**Comment by iRipVanWinkle** [comment]
@klimov-paul #12862 it issue only Migration, but my issue about all console commands.

**Comment by cebe** [comment]
> if need I can implement it?

sure, if you want to work on it, pull request is welcome :)

> #12862 it issue only Migration, but my issue about all console commands.

it is only related, not duplicate :)

**Comment by TerraSkye** [comment]
I can see it a use case. When i run the command myself i wanna see the output. But for cronjob i dont want it

**Comment by cebe** [comment]
@TerraSkye for a cronjob you can just pipe the output to `/dev/null`, so no change is needed for that use case.

**Comment by SamMousa** [comment]
The original problem here is the fact executing a migration pushes many people to instantiate the console controller.
Ideally, the migrations should be runnable via some migration service / component that is used by the controller.
That way if you want to run migrations from the web (I think that's what you're doing) you can just talk to the component / service instead of instantiating the controller and running an action on it.

**Comment by samdark** [comment]
@SamMousa the problem you've mentioned is valid but it's a separate problem which serves as an example.

**Comment by SamMousa** [comment]
Yes, but logic shouldn't be in the controller layer right?

**Comment by samdark** [comment]
Ideally yes.

**Comment by SamMousa** [comment]
So that part we could fix for migrations (since it is part of the core). 
I don't think the argument for the silent option should be to use it via another (web) controller.
On the other hand if the argument is use cases like cron which send mail if there's output; then that makes sense 👍

**Comment by samdark** [comment]
@SamMousa https://github.com/yiisoft/yii2/issues/12862

**Comment by samdark** [comment]
Won't be implemented in 2.0.

**Comment by martingold** [comment]
For anyone wondering how to suppress output as @iRipVanWinkle suggested:

```php
        $options = [$options, ...[
            'interactive' => 0,
            'compact' => true,
        ]];

        if ($this->verbose) {
            Yii::$app->runAction($action, $options);
            return;
        }

        // Register filter to suppress all command output
        $discardOutputFilter = new class extends php_user_filter
        {
            public string $filtername = 'discard_output';

            /**
             * @param resource $in
             * @param resource $out
             * @param int $consumed
             */
            public function filter($in, $out, &$consumed, bool $closing): int
            {
                while ($bucket = stream_bucket_make_writeable($in)) {
                    $bucket->data = '';
                    $consumed += $bucket->datalen;
                    stream_bucket_append($out, $bucket);
                }
                return PSFS_PASS_ON;
            }
        };

        stream_filter_register('discard_output', get_class($discardOutputFilter));
        $filter = stream_filter_append(STDOUT, 'discard_output');

        Yii::$app->runAction($action, $options);

        stream_filter_remove($filter);
```
